mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp

```diff
@@ -66,8 +69,125 @@ void strToTokens(const std::string &arguments,
+      config.strideWidth};
+  if (std::any_of(strictlyPositiveParams.begin(), strictlyPositiveParams.end(),
+                  [](const int64_t &a) { return a <= 0; })) {
+    llvm::errs() << "Dilation and stride must be a positive integer\n";
+    return failure();
+  }
+
+  static const SmallVector<int64_t, 4> nonNegativeParams{
+      config.paddingHeightLeft, config.paddingHeightRight,
+      config.paddingWidthLeft, config.paddingWidthRight};
+  if (std::any_of(nonNegativeParams.begin(), nonNegativeParams.end(),
+                  [](const int64_t &a) { return a < 0; })) {
+    llvm::errs() << "Padding values cannot be negative\n";
+    return failure();
+  }
+
+  auto checkDimSizes = [](const SmallVector<int64_t, 5> &dims) -> bool {
+    return std::all_of(dims.begin(), dims.end(),
+                       [](const int64_t &a) { return a > 0; });
+  };
+
+  if (!checkDimSizes(config.inputDimension)) {
+    llvm::errs() << "Input tensor dimensions must be strictly positive\n";
+    return failure();
+  }
+  if (!checkDimSizes(config.filterDimension)) {
+    llvm::errs() << "Filter tensoor dimensions must be strictly positive\n";
+  }
+  if (!checkDimSizes(config.outputDimension)) {
+    llvm::errs() << "Output tensor dimensions must be strictly positive\n";
+    return failure();
+  }
+
+  auto inDimMaybe =
+      canonicalizeDims(config.inputDimension, config.inputLayout, "ngchw");
+  auto filDimMaybe =
+      canonicalizeDims(config.filterDimension, config.filterLayout, "gkcyx");
+  auto outDimMaybe =
+      canonicalizeDims(config.outputDimension, config.outputLayout, "ngkhw");
+  if (!(inDimMaybe.hasValue() && filDimMaybe.hasValue() &&
+        outDimMaybe.hasValue())) {
+    llvm_unreachable("Tensor dimensions should have been checked before now in "
+                     "isValidConvolution()!");
+  }
+  auto inDim = inDimMaybe.getValue();
+  auto filDim = filDimMaybe.getValue();
+  auto outDim = outDimMaybe.getValue();
+
+  if (inDim[0] != outDim[0]) {
+    llvm::errs() << "Input and output batch sizes don't match\n";
+    return failure();
+  }
+  if (inDim[1] != outDim[1] || inDim[1] != filDim[0]) {
+    llvm::errs()
+        << "Group sizes are not consistent between input, output, and filter\n";
+    return failure();
+  }
+  if (inDim[2] != filDim[2]) {
+    llvm::errs() << "Number of channels in input doesn't match number of "
+                    "channels in filter\n";
+    return failure();
+  }
+  if (filDim[1] != outDim[2]) {
+    llvm::errs() << "Number of channels in output doesn't match number of "
+                    "channels in filter\n";
+    return failure();
+  }
+
+  int64_t expectedOutHeight = outputDim(
+      inDim[3], filDim[3], config.paddingHeightLeft, config.paddingHeightRight,
+      config.strideHeight, config.dilationHeight);
+  int64_t expectedOutWidth = outputDim(
+      inDim[4], filDim[4], config.paddingWidthLeft, config.paddingWidthRight,
+      config.strideWidth, config.dilationWidth);
```

Comment:
If I understand it right, you will only set output dimensions when the client did not pass them in. When client pass in a output dimension, there's a possibility that it is wrong, therefore the need for a check?

Reply:
I believe I did add such a check.

Reply:
Yes you have. I forget to add the `[Question]` tag in the previous comment. But basically I was thinking loud and try to answer myself in why output weight/height was already calculated and checked again - The answer seems to be that at certain cases we will not do the original calculation to figure out what output weight/height was therefore the need to check.